### ISSUE TITLE
Unhandled Internal Exception upon rule with reflexive relation

### ISSUE THREAD

**Issue by JRWest2000** [vaticle/typedb#6359]
~~I Know that you do not support recursive rules but I created this bug report because I am sure that crashing is not the way you want to handle recursive quires.~~ (see https://github.com/vaticle/typedb/issues/6359#issuecomment-856589173)

While running a complex 200+ rule query with recursion the console gave this Java error  

java.lang.UnsupportedOperationException Please check server logs for the stack trace.

Note: the files uploaded are not indented to be a realistic project they are indented to explore the possibility that there are problems ware the reasoning is so complex it can only be solved using a tool like TypeDB

Stand Alone PC
OS Windows 10
TypeDB 2.1.3
running *.gql files thru the console using the source command

create empty database
do transaction test schema write & load schema file
do transaction test data write & load definition file
do transaction test data read & load solver file

## Expected Output

Some kind of traceable error that would let me know which recursive rule triggered the problem 

## Actual Output

An untraded Java exception that makes it impossible to write the quires needed to trace the logic chain that lead to the crash
 
## Additional information

I am including my *.gql  files total run time to load and run to error < 5 min.
the ERROR file is the complete text of the server cmd window output

(had to rename them *.txt the .gql is not supported)

[Math-Schema.txt](https://github.com/vaticle/typedb/files/6613805/Math-Schema.txt)
[Definition.txt](https://github.com/vaticle/typedb/files/6613830/Definition.txt)
[Solver.txt](https://github.com/vaticle/typedb/files/6613836/Solver.txt)
[ERROR.txt](https://github.com/vaticle/typedb/files/6613899/ERROR.txt)

**Comment by jmsfltchr** [comment]
Thanks for reporting @JRWest2000! I would like to correct you, that TypeDB absolutely can do recursive reasoning! The error is actually very specific in this case, so I'll look into it.

As a point of interest, did you hand-write the rules or generate them? Writing a calculator in TypeDB is very cool, I haven't seen that before.

**Comment by JRWest2000** [comment]
Not a calculator. An algebraic solver that extends to requirements like

divided by c --> c can not be zero, 
greater than --> not equal to, 
a + b both positive --> must be greater than a and b 

as you can see these kinds of inferences can number into the thousands or even millions when you start working with things like fluid mechanics or particle physics and could give amazing insights into the strengths and weakness of modeling techniques

**Comment by jmsfltchr** [comment]
Noted, I saw that there was a lot of algebra described over operators and made the leap to calculator.

I wanted to give you an update. I've done some debugging and found that the problem centres around the pattern of reflexive relations: relations playing a role in themselves. In the example below `$d` is the relation and also plays the role `clc-rhs` in itself. In this context I'm not sure that makes sense, but I'll leave it to you to determine that. There are many rules that exhibit this pattern in the schema you shared.
```
# b - c && a - c && (a = b | b = c)
# --> b - c = a - c
# ==> a - c = b - c
rule subtraction-equality-substiutions-are-equal:
when {
      $d (clc-lhs: $a, clc-rhs: $d) isa subtraction;
      $e (clc-lhs: $b, clc-rhs: $d) isa subtraction;
      (cmp-lhs: $a, cmp-rhs: $b) isa Equality;
     }
then {
      (cmp-lhs: $d, cmp-rhs: $e) isa Equality;
     };
```
Hope that helps while I'm looking to fix this.

**Comment by JRWest2000** [comment]
That pattern is common in modeling any binary tree  'a' is the node representing 'b - c' so you represent 'd - b -c' as 'd - a'

**Comment by JRWest2000** [comment]
@jmsfltchr 
You are correct it should be. ...

\# b - c && a - c && (a = b | b = c)
\# --> b - c = a - c
\# ==> a - c = b - c
rule subtraction-equality-substiutions-are-equal:
when {
      $d (clc-lhs: $a, clc-rhs: $c) isa subtraction;
      $e (clc-lhs: $b, clc-rhs: $c) isa subtraction;
      (cmp-lhs: $a, cmp-rhs: $d) isa Equality;
     }
then {
      (cmp-lhs: $d, cmp-rhs: $e) isa Equality;
     };

I was in the middle of something when I saw your reply and missed it

**Comment by jmsfltchr** [comment]
That's good news as you should be unblocked from this issue! You have many rules where this occurs so please correct them also (if indeed they are also unintentionally reflexive).
I've made progress by fixing the exact error you were seeing in https://github.com/vaticle/typedb/pull/6366. Once they were gone a deeper issue was revealed, you or anyone finding themselves here can track that issue https://github.com/vaticle/typedb/issues/6367. When that is solved we expect this issue will also be solved.

**Comment by flyingsilverfin** [comment]
Closing because as you are unblocked with your issue and we have resolved this into #6366  and #6367
